### DISCUSSION TITLE
Refresh token invalidated when using a previous (revoked) one

### DISCUSSION THREAD

**Discussion by de7erv** [keycloak/keycloak#11318] [Q&A]
Hi all,

With a client, we have the following situation:
 - Client requests new refresh/access token > OK > client receives new refresh and access token
 - Client erroneously requests a new access token with the old revoked refresh token x3 > Error - as designed normal behavior
 - Client uses correctly the last generated refresh token that is valid for his user > NOK - same error as above.
 
 It seems the logic is that if the client uses a wrong refresh token, all refresh tokens for that user are invalidated. 
 
 My question is: Can we change this in the config or is it not possible to change this behaviour?

**Comment by soylent-grin** [comment]
Hi @de7erv,

I encountered same issue, the problem is that user session is detached on disallowed refresh token re-use: https://github.com/keycloak/keycloak/blame/62fd969fe1b8630f093c6ff7f4bc4c0f302b54d0/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L452
But this session is common for all refresh tokens (issued during current session), so consequent valid refresh tokens lose their sessions too.
This is recommended behaviour according to OAuth:
```
If the particular
   token is a refresh token and the authorization server supports the
   revocation of access tokens, then the authorization server SHOULD
   also invalidate all access tokens based on the same authorization
   grant (see Implementation Note).  If the token passed to the request
   is an access token, the server MAY revoke the respective refresh
   token as well.
```
Seems that the only workaround is to increase max refresh token reuse (which sounds equivalent to disable token rotation)
